### PR TITLE
TEST: test reduction operations

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
## What
Add reduction tests to gtest

### CODE REVIEW

File: test/gtest/core/test_mc_reduce.cc [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                                           ReductionTest<UCC_DT_FLOAT64, max>,
+                                           ReductionTest<UCC_DT_FLOAT32, min>,
+                                           ReductionTest<UCC_DT_FLOAT64, min>,
+                                           ReductionTest<UCC_DT_FLOAT32, sum>,
+                                           ReductionTest<UCC_DT_FLOAT64, sum>,
+                                           ReductionTest<UCC_DT_FLOAT32, prod>,
+                                           ReductionTest<UCC_DT_FLOAT64, prod>>;
+
+TYPED_TEST_CASE(test_mc_reduce, ReductionTypesOps);
+TYPED_TEST(test_mc_reduce, ucc_op_sum) {
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce.cc:249]: why "ucc_op_sum" ? it covers all combinations right?

File: test/gtest/core/test_mc_reduce.cc [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        ASSERT_EQ(arg1, arg2);
+    }
+    static type do_op(type arg1, type arg2) {
+        op<type> _op;
+        return _op(arg1, arg2);
+    }
+};
+
+template<template <typename P> class op>
+struct ReductionTest<UCC_DT_INT64, op>
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce.cc:45]: Maybe make a macro to declare this template class: the code is identical almost entirely:
#define REDUCTION_TEST_DECLARE(_type, _ucc_type, _assert_eq) do {
using type = _type; \
const status ucc_datatype_t dt = _ucc_type; \
ucc_reduction_op redop = op<type>::redop;\
static void assert_equal(...) { \
_assert_eq(arg1, arg2); \
} \
...
} while (0)

and then 
REDUCTION_TEST_DECLARE(int64, UCC_DT_INT64, ASSERT_EQ);
REDUCTION_TEST_DECLARE(float, UCC_DT_FLOAT32, ASSERT_EQ_FLOAT);

etc

File: test/gtest/core/test_mc_reduce.cc [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+                                           ReductionTest<UCC_DT_FLOAT32, sum>,
+                                           ReductionTest<UCC_DT_FLOAT64, sum>,
+                                           ReductionTest<UCC_DT_FLOAT32, prod>,
+                                           ReductionTest<UCC_DT_FLOAT64, prod>>;
+
+TYPED_TEST_CASE(test_mc_reduce, ReductionTypesOps);
+TYPED_TEST(test_mc_reduce, ucc_op_sum) {
+    ucc_mc_reduce(this->buf1, this->buf2, this->res,
+                  this->COUNT, TypeParam::dt,
+                  TypeParam::redop, UCC_MEMORY_TYPE_HOST);
```
Comment (vspetrov) [inline test/gtest/core/test_mc_reduce.cc:252]: what about MEMORY_TYPE_CUDA?

Comment (vspetrov) [reply test/gtest/core/test_mc_reduce.cc:252]: What about reduce_multi?

**Review by vspetrov — APPROVED** [review]

File: test/gtest/core/test_mc_reduce.cc (hunk 2 above)
Comment (Sergei-Lebedev) [reply test/gtest/core/test_mc_reduce.cc:45]: Sure, it's just not easy to debug when test is wrapped into macro. But I want first extend test with memory type and add reduce multi.

File: test/gtest/core/test_mc_reduce.cc (hunk 3 above)
Comment (Sergei-Lebedev) [reply test/gtest/core/test_mc_reduce.cc:252]: Yes, if overall approach looks good I will add cuda and reduce multi

File: test/gtest/core/test_mc_reduce.cc (hunk 2 above)
Comment (vspetrov) [reply test/gtest/core/test_mc_reduce.cc:45]: we can keep it as is if you think it makes inconvenient debug

Comment (Sergei-Lebedev) [reply test/gtest/core/test_mc_reduce.cc:45]: i mean debug of test itself, once all tests parameters are there (reduce, reduce multi, mem types) we can wrap class declaration in macro

**Comment by Sergei-Lebedev** [comment]
@vspetrov added cuda and reduce multi, pls check

**Comment by vspetrov** [comment]
@manjugv @bureddy plz have a look. It's been ready for a while

**Review by bureddy — APPROVED** [review]

**Pull request merged**